sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
+        _all.insert(sst);
+        _times_added[sst] = 1;
+    } catch (...) {
+        _impl->erase(sst);
+        if (_all_runs.contains(sst->run_identifier())) {
+            _all_runs.at(sst->run_identifier()).erase(sst);
+        }
+        _all.erase(sst);
+        _times_added.erase(sst);
+        throw;
+    }
+}
+
+std::vector<shared_sstable> versioned_sstable_set_data::select(const dht::partition_range& range) const {
+    return _impl->select(range);
+}
+
+std::unordered_set<shared_sstable> versioned_sstable_set_data::select_by_run_id(utils::UUID run_id) const {
+    return _all_runs.at(run_id);
 }
 
-sstable_set::sstable_set(sstable_set&&) noexcept = default;
+const std::set<shared_sstable>& versioned_sstable_set_data::all() const {
+    return _all;
+}
 
-sstable_set&
-sstable_set::operator=(const sstable_set& x) {
-    if (this != &x) {
-        auto tmp = sstable_set(x);
-        *this = std::move(tmp);
+void versioned_sstable_set_data::remove(shared_sstable sst) {
+    if (--_times_added[sst] == 0) {
+        _impl->erase(sst);
+        _all_runs[sst->run_identifier()].erase(sst);
+        _all.erase(sst);
+        _times_added.erase(sst);
     }
-    return *this;
 }
 
-sstable_set&
-sstable_set::operator=(sstable_set&&) noexcept = default;
+std::unique_ptr<incremental_selector_impl> versioned_sstable_set_data::make_incremental_selector() {
+    return _impl->make_incremental_selector();
+}
 
-std::vector<shared_sstable>
-sstable_set::select(const dht::partition_range& range) const {
-    return _impl->select(range);
+// version operations
+sstable_set_version_reference::~sstable_set_version_reference() {
+    if (_p) {
+        _p->_reference_count--;
+        if (_p->_reference_count == 1 && _p->_next.size() == 1) {
+            _p->merge_with_next();
+        } else if (_p->_reference_count == 0) {
+            delete _p;
+            _p = nullptr;
+        }
+    }
 }
 
-std::vector<sstable_run>
-sstable_set::select_sstable_runs(const std::vector<shared_sstable>& sstables) const {
-    auto run_ids = boost::copy_range<std::unordered_set<utils::UUID>>(sstables | boost::adaptors::transformed(std::mem_fn(&sstable::run_identifier)));
-    return boost::copy_range<std::vector<sstable_run>>(run_ids | boost::adaptors::transformed([this] (utils::UUID run_id) {
-        return _all_runs.at(run_id);
-    }));
+sstable_set_version_reference::sstable_set_version_reference(sstable_set_version* p) : _p(p) {
+    if (_p) {
+        _p->_reference_count++;
+    }
 }
 
-void
-sstable_set::insert(shared_sstable sst) {
-    _impl->insert(sst);
-    try {
-        _all->insert(sst);
-        try {
-            _all_runs[sst->run_identifier()].insert(sst);
-        } catch (...) {
-            _all->erase(sst);
-            throw;
+sstable_set_version_reference::sstable_set_version_reference(const sstable_set_version_reference& ref) 
+    : _p(ref._p ? new sstable_set_version(ref._p) : nullptr) {
```

Comment:
Why isn't this increasing the reference count only?

It's inconsistent that copying a reference creates a logical clone.

Reply:
We don't actually ever want to copy an `sstable_set_version_reference`, but when we copy a class that has a sstable_set_version_reference (i.e. an sstable_list or sstable_set), we want to have a reference to a new version in it, so I used the copy constructor for it.
I can see though that it's not really consistent, so I can change the copy constructor to be marked deleted and use a custom constructor instead, if that was simpler